### ISSUE TITLE
Build crash if source code contains a nested record type and apply obfuscate.

### ISSUE THREAD

**Issue by lamtx** [flutter/flutter#125289]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

If the source code contains a nested record type, the compiler crash when building if option `--obfuscate`.
There are no issues when building debug.

Steps to reproduce:
1. Create a new project: `flutter create example`.
2. Replace file main.dart by the attached source code.
3. Build APK or IPA by following command:
`flutter build apk --obfuscate --split-debug-info=./split-debug-info`
or 
`flutter build ipa --obfuscate --split-debug-info=./split-debug-info`

### Expected results

Build successfully.

### Actual results

The compiler crashes.

### Code sample

<details><summary>main.dart</summary>

```dart
import 'package:flutter/material.dart';

const config = (appbar: (title: "Example"));

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: Text(config.appbar.title),
        ),
      ),
    );
  }
}

```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details><summary>Logs</summary>

```console
may@Mays-Mac-mini example % flutter build apk --obfuscate --split-debug-info=./split-debug-info

Font asset "MaterialIcons-Regular.otf" was tree-shaken, reducing it from 1645184 to 1284 bytes (99.9% reduction). Tree-shaking can be disabled by providing the --no-tree-shake-icons flag when building your app.

===== CRASH =====
si_signo=Segmentation fault: 11(11), si_code=1, si_addr=0x11
version=3.0.0-417.2.beta (beta) (Wed Apr 19 09:44:49 2023 +0000) on "macos_arm"
pid=6825, thread=-1, isolate_group=isolate(0x7fd87b808200), isolate=(nil)(0x0)
os=macos, arch=arm, comp=no, sim=no
isolate_instructions=0, vm_instructions=0
fp=30d159ea0, sp=30d159e70, pc=10443cd10
  pc 0x000000010443cd10 fp 0x000000030d159ea0 dart::AbstractType::IsStrictlyNonNullable() const+0x30
  pc 0x00000001046d9424 fp 0x000000030d159f00 dart::LoadFieldInstr::ComputeType() const+0x264
  pc 0x00000001046d0a34 fp 0x000000030d15a380 dart::FlowGraphTypePropagator::VisitValue(dart::Value*)+0xf4
  pc 0x00000001046d0577 fp 0x000000030d15a410 dart::FlowGraphTypePropagator::PropagateRecursive(dart::BlockEntryInstr*)+0x677
  pc 0x00000001046d0673 fp 0x000000030d15a4a0 dart::FlowGraphTypePropagator::PropagateRecursive(dart::BlockEntryInstr*)+0x773
  pc 0x00000001046fb1cd fp 0x000000030d15aa20 dart::CompilerPass_TypePropagation::DoBody(dart::CompilerPassState*) const+0x96d
  pc 0x00000001046fef07 fp 0x000000030d15aaf0 dart::CompilerPass::Run(dart::CompilerPassState*) const+0x127
  pc 0x00000001046ff4df fp 0x000000030d15ab20 dart::CompilerPass::RunPipeline(dart::CompilerPass::PipelineMode, dart::CompilerPassState*)+0x4f
  pc 0x00000001045b12ce fp 0x000000030d15b1f0 dart::PrecompileParsedFunctionHelper::Compile(dart::CompilationPipeline*)+0xabe
  pc 0x00000001045b0533 fp 0x000000030d15b880 dart::PrecompileFunctionHelper(dart::Precompiler*, dart::CompilationPipeline*, dart::Function const&, bool)+0x5c3
  pc 0x00000001045ae887 fp 0x000000030d15b990 dart::Precompiler::CompileFunction(dart::Precompiler*, dart::Thread*, dart::Zone*, dart::Function const&)+0x1e7
  pc 0x00000001045967a4 fp 0x000000030d15c740 dart::Precompiler::DoCompileAll()+0xa284
  pc 0x000000010458c0a9 fp 0x000000030d15cbc0 dart::Precompiler::CompileAll()+0x979
  pc 0x00000001047e01af fp 0x000000030d15cd30 Dart_Precompile+0x2cf
  pc 0x00000001042b4285 fp 0x000000030d15d280 dart::bin::main(int, char**)+0x15e5
  pc 0x0000000204db441f fp 0x000000030d15d510 Unknown symbol
-- End of DumpStackTrace
=== Crash occurred when compiling package:example/main.dart_Po_ewa in AOT mode in TypePropagation pass
*** BEGIN CFG
TypePropagation
==== package:example/main.dart_Po_ewa (RegularFunction)
B0[graph]:0 {
      v0 <- Constant(#null)
      v1 <- Constant(#<optimized out>)
      v4 <- Constant(#Record (Kvc: Record (IYa: Example))) T{_Record}
}
B1[function entry]:2 {
      v2 <- Parameter(0)
      v3 <- Parameter(1)
}
    CheckStackOverflow:8(stack=0, loop=0)
    v5 <- LoadField(v4 . :record_field)
    v6 <- LoadField(v5 . :record_field)
    v7 <- AllocateObject:10(cls=QR) T{QR}
    StaticCall:12( QR.<0> v7)
    v8 <- AllocateObject:14(cls=vw) T{vw}
    StaticCall:16( vw.<0> v8, v7)
    v9 <- AllocateObject:18(cls=Dx) T{Dx}
    StaticCall:20( Dx.<0> v9, v8)
    v10 <- AllocateObject:22(cls=ow) T{ow}
    StaticCall:24( ow.<0> v10, v9)
    Return:26(v10)
*** END CFG
Dart snapshot generator failed with exit code -6
Target android_aot_release_android-arm failed: Exception: AOT snapshotter exited with code -6


FAILURE: Build failed with an exception.

* Where:
Script '/Users/may/Apps/flutter/packages/flutter_tools/gradle/flutter.gradle' line: 1159

* What went wrong:
Execution failed for task ':app:compileFlutterBuildRelease'.
> Process 'command '/Users/may/Apps/flutter/bin/flutter'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 12s
Running Gradle task 'assembleRelease'...                           13.4s
Gradle task assembleRelease failed with exit code 1

```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
may@Mays-Mac-mini ~ % flutter doctor -v
[✓] Flutter (Channel beta, 3.10.0-1.3.pre, on macOS 13.3 22E252 darwin-arm64,
    locale en-VN)
    • Flutter version 3.10.0-1.3.pre on channel beta at /Users/may/Apps/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision d11aff97d2 (2 days ago), 2023-04-19 11:11:05 -0700
    • Engine revision a645b36a9b
    • Dart version 3.0.0 (build 3.0.0-417.2.beta)
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
    • Android SDK at /Users/may/Library/Android/sdk
    • Platform android-33, build-tools 33.0.2
    • Java binary at: /Applications/Android
      Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      11.0.15+0-b2043.56-8887301)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14C18
    • CocoaPods version 1.12.0

[✗] Chrome - develop for the web (Cannot find Chrome executable at
    /Applications/Google Chrome.app/Contents/MacOS/Google Chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[✓] Android Studio (version 2022.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      11.0.15+0-b2043.56-8887301)

[✓] Connected device (2 available)
    • iPhone 14 Pro Max (mobile) • 7D5E09BD-FE97-42B4-9DF5-DE35A11D8994 • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-16-2 (simulator)
    • macOS (desktop)            • macos                                •
      darwin-arm64 • macOS 13.3 22E252 darwin-arm64

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

</details>

**Comment by danagbemava-nc** [comment]
The crash is reproducible using the code sample and the steps provided above. 

As reported, the crash only occurs when the `--obfuscate` flag is added. 

Logs from build ipa -v have been shared below. It looks almost the same as the android crash logs shared above. 

<details>
<summary>iOS crash logs (taken from verbose output)</summary>

```
===== CRASH =====
                     [        ] si_signo=Segmentation fault: 11(11), si_code=1, si_addr=0x11
                     [        ] version=3.1.0-26.0.dev (dev) (Thu Apr 20 12:25:58 2023 -0700) on "macos_simarm64"
                     [        ] pid=26479, thread=-1, isolate_group=isolate(0x7fc770808200), isolate=(nil)(0x0)
                     [        ] os=macos, arch=arm64, comp=no, sim=yes
                     [        ] isolate_instructions=0, vm_instructions=0
                     [        ] fp=30ea02170, sp=30ea02140, pc=105084d50
                     [   +1 ms]   pc 0x0000000105084d50 fp 0x000000030ea02170 dart::AbstractType::IsStrictlyNonNullable() const+0x30
                     [        ]   pc 0x0000000105328904 fp 0x000000030ea021d0 dart::LoadFieldInstr::ComputeType() const+0x264
                     [        ]   pc 0x000000010531ff24 fp 0x000000030ea02650 dart::FlowGraphTypePropagator::VisitValue(dart::Value*)+0xf4
                     [        ]   pc 0x000000010531fa67 fp 0x000000030ea026e0 dart::FlowGraphTypePropagator::PropagateRecursive(dart::BlockEntryInstr*)+0x677
                     [        ]   pc 0x000000010531fb63 fp 0x000000030ea02770 dart::FlowGraphTypePropagator::PropagateRecursive(dart::BlockEntryInstr*)+0x773
                     [        ]   pc 0x00000001053488bd fp 0x000000030ea02cf0 dart::CompilerPass_TypePropagation::DoBody(dart::CompilerPassState*) const+0x96d
                     [        ]   pc 0x000000010534c7b7 fp 0x000000030ea02dc0 dart::CompilerPass::Run(dart::CompilerPassState*) const+0x127
                     [        ]   pc 0x000000010534cd8f fp 0x000000030ea02df0 dart::CompilerPass::RunPipeline(dart::CompilerPass::PipelineMode, dart::CompilerPassState*)+0x4f
                     [        ]   pc 0x0000000105207bfe fp 0x000000030ea034c0 dart::PrecompileParsedFunctionHelper::Compile(dart::CompilationPipeline*)+0xabe
                     [        ]   pc 0x0000000105206e63 fp 0x000000030ea03b50 dart::PrecompileFunctionHelper(dart::Precompiler*, dart::CompilationPipeline*, dart::Function const&, bool)+0x5c3
                     [        ]   pc 0x00000001052052b7 fp 0x000000030ea03c60 dart::Precompiler::CompileFunction(dart::Precompiler*, dart::Thread*, dart::Zone*, dart::Function const&)+0x1e7
                     [        ]   pc 0x00000001051edae5 fp 0x000000030ea049b0 dart::Precompiler::DoCompileAll()+0xa205
                     [        ]   pc 0x00000001051e3469 fp 0x000000030ea04e30 dart::Precompiler::CompileAll()+0x979
                     [        ]   pc 0x0000000105430f8f fp 0x000000030ea04fa0 Dart_Precompile+0x2cf
                     [        ]   pc 0x0000000104efd285 fp 0x000000030ea054f0 dart::bin::main(int, char**)+0x15e5
                     [        ]   pc 0x0000000205a1d41f fp 0x000000030ea05780 Unknown symbol
                     [        ] -- End of DumpStackTrace
                     [        ] === Crash occurred when compiling package:mcrash/main.dart_US_lwa in AOT mode in TypePropagation pass
                     [        ] *** BEGIN CFG
                     [        ] TypePropagation
                     [        ] ==== package:mcrash/main.dart_US_lwa (RegularFunction)
                     [        ] B0[graph]:0 {
                     [        ]       v0 <- Constant(#null)
                     [        ]       v1 <- Constant(#<optimized out>)
                     [        ]       v4 <- Constant(#Record (dwc: Record (LWa: Example))) T{_Record}
                     [        ] }
                     [        ] B1[function entry]:2 {
                     [        ]       v2 <- Parameter(0)
                     [        ]       v3 <- Parameter(1)
                     [        ] }
                     [        ]     CheckStackOverflow:8(stack=0, loop=0)
                     [        ]     v5 <- LoadField(v4 . :record_field)
                     [        ]     v6 <- LoadField(v5 . :record_field)
                     [   +2 ms]     v7 <- AllocateObject:10(cls=TR) T{TR}
                     [        ]     StaticCall:12( TR.<0> v7)
                     [        ]     v8 <- AllocateObject:14(cls=ww) T{ww}
                     [        ]     StaticCall:16( ww.<0> v8, v7)
                     [        ]     v9 <- AllocateObject:18(cls=Ex) T{Ex}
                     [        ]     StaticCall:20( Ex.<0> v9, v8)
                     [        ]     v10 <- AllocateObject:22(cls=pw) T{pw}
                     [        ]     StaticCall:24( pw.<0> v10, v9)
                     [        ]     Return:26(v10)
                     [        ] *** END CFG
                     [        ] Dart snapshot generator failed with exit code -6
                     [   +4 ms] Target aot_assembly_release failed: Exception: AOT snapshotter exited with code -6
```
</details>

<details>
<summary>flutter doctor -v</summary>

```
[!] Flutter (Channel beta, 3.10.0-1.3.pre, on macOS 13.3.1 22E261 darwin-arm64, locale en-GB)
    • Flutter version 3.10.0-1.3.pre on channel beta at /Users/nexus/dev/sdks/flutter_beta
    ! Warning: `flutter` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutter_beta. Consider adding /Users/nexus/dev/sdks/flutter_beta/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutter_beta. Consider adding /Users/nexus/dev/sdks/flutter_beta/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision d11aff97d2 (2 days ago), 2023-04-19 11:11:05 -0700
    • Engine revision a645b36a9b
    • Dart version 3.0.0 (build 3.0.0-417.2.beta)
    • DevTools version 2.23.1
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.9862592/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode-14.3.0.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.1)
    • Android Studio at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/221.6008.13.2211.9619390/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)

[✓] Android Studio (version 2022.2)
    • Android Studio at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.9862592/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.1)
    • IntelliJ at /Users/nexus/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 73.0.4
    • Dart plugin version 231.8109.91

[✓] IntelliJ IDEA Ultimate Edition (version 2023.1)
    • IntelliJ at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/231.8109.175/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.77.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.62.0

[✓] Connected device (3 available)
    • Nexus (mobile)  • 00008020-001875E83A38002E • ios            • iOS 16.4.1 20E252
    • macOS (desktop) • macos                     • darwin-arm64   • macOS 13.3.1 22E261 darwin-arm64
    • Chrome (web)    • chrome                    • web-javascript • Google Chrome 112.0.5615.137
    ! Error: Nexus is busy: Fetching debug symbols for Nexus. Xcode will continue when Nexus is finished. (code -10)

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

```
[!] Flutter (Channel master, 3.10.0-10.0.pre.39, on macOS 13.3.1 22E261 darwin-arm64, locale en-GB)
    • Flutter version 3.10.0-10.0.pre.39 on channel master at /Users/nexus/dev/sdks/flutters
    ! Warning: `flutter` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/nexus/dev/sdks/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/nexus/dev/sdks/flutters. Consider adding /Users/nexus/dev/sdks/flutters/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision d186792c00 (6 hours ago), 2023-04-20 20:50:59 -0700
    • Engine revision 122c3b3820
    • Dart version 3.1.0 (build 3.1.0-26.0.dev)
    • DevTools version 2.23.1
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/nexus/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.9862592/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Applications/Xcode-14.3.0.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.1)
    • Android Studio at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/221.6008.13.2211.9619390/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)

[✓] Android Studio (version 2022.2)
    • Android Studio at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.9862592/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.1)
    • IntelliJ at /Users/nexus/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 73.0.4
    • Dart plugin version 231.8109.91

[✓] IntelliJ IDEA Ultimate Edition (version 2023.1)
    • IntelliJ at /Users/nexus/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/231.8109.175/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.77.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.62.0

[✓] Connected device (3 available)
    • Nexus (mobile)  • 00008020-001875E83A38002E • ios            • iOS 16.4.1 20E252
    • macOS (desktop) • macos                     • darwin-arm64   • macOS 13.3.1 22E261 darwin-arm64
    • Chrome (web)    • chrome                    • web-javascript • Google Chrome 112.0.5615.137

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```
</details>

**Comment by mraleph** [comment]
Probably related to https://github.com/dart-lang/sdk/issues/52124

/cc @alexmarkov @mkustermann

**Comment by alexmarkov** [comment]
The fix rolled into Flutter in 6382136e1f45cd08c14b874081a4e5537dbf09b5.
